### PR TITLE
 Check if maxpagesize should be reinjected 

### PR DESCRIPTION
**Pull request by radhgupta**
This pull request updates the logic for appending query parameters in the generated REST client code to ensure that duplicate query parameters are not added when handling "next link" requests (typically used for pagination). The changes introduce runtime checks to prevent reinjecting parameters that are already present in the query string, and update the relevant test cases to reflect this behavior.
Fixes: https://github.com/microsoft/typespec/issues/9201

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestNextLinkReinjectedParametersInCreateRequestMethod.cs [hunk 1]
```diff
@@ -27,8 +27,14 @@ public partial class TestClient
         {
             global::Sample.ClientUriBuilder uri = new global::Sample.ClientUriBuilder();
             uri.Reset(nextPage);
-            uri.AppendQuery("someOtherName", p1, true);
-            uri.AppendQuery("p2", p2, true);
+            if (!uri.Query.Contains("someOtherName="))
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestNextLinkReinjectedParametersInCreateRequestMethod.cs]: I think we should probably start off with just scoping this change to the `maxpagesize` parameter. Have we ran into another instance of a server rejecting the payload due to another duplicate parameter? If it's needed in the future, we should be able to easily rescope this to all reinjected params.

Comment (radhgupta) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestNextLinkReinjectedParametersInCreateRequestMethod.cs]: I haven't seen any other instance in GitHub issues related to duplicate parameter. I assumed it would be same with other parameters as well. Thanks, I can make the changes specific to maxpagesize

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs [hunk 2]
```diff
@@ -34,7 +34,10 @@ public partial class TestClient
             uri.AppendQuery("p1", p1, true);
             if ((maxPageSize != null))
             {
-                uri.AppendQuery("maxPageSize", global::Sample.TypeFormatters.ConvertToString(maxPageSize), true);
+                if (!uri.Query.Contains("maxPageSize="))
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs]: I think we need to handle updating the maxPageSize parameter value. The value in the nextLink might need to be replaced by the value passed to this method. So this is essentially updating the query param if it already exists, else appending.

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs]: Side note, is `Query` an accessible property of the uribuilder type?

Comment (radhgupta) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs]: Thanks for letting me know, I assumed the bug wrong! I have added the logic to rather update the parameter value. I have made sure to keep the logic specific to maxpagesize but also ensured that we can add more parameter in future.

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs [hunk 3]
```diff
@@ -34,7 +34,20 @@ public partial class TestClient
             uri.AppendQuery("p1", p1, true);
             if ((maxPageSize != null))
             {
-                if (!uri.Query.Contains("maxPageSize="))
+                if (uri.Query.Contains("maxPageSize="))
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs]: I think there is still an issue here. `Query` isn't an accessible property in the ClientUriBuilder class, unless I'm missing something here, so this code won't compile. In general, when it comes to writing code for a generator, I highly recommend handwriting the changes first to get the implementation down and then adding the code to get it generated. In this case, I think we should use the [ClientUriBuilderSnippets](https://github.com/microsoft/typespec/blob/0e3d0b9d02afd1d3a4695351c7bedcf782fe75b7/packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Snippets/ClientUriBuilderSnippets.cs#L13) to construct the condition. It is okay if we have to add/transform a generated property in that class for the sake of supporting this.

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs [hunk 4]
```diff
@@ -34,7 +34,20 @@ public partial class TestClient
...
+                {
+                    string currentQuery = uri.Query;
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs]: I'm wondering if this should be a helper method in ClientUriBuilder? Something like "UpdateQuery"? It would also enable us to add a unit test for this query updating logic.

Comment (radhgupta) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/RestClientProviders/TestData/RestClientProviderTests/TestPageSizeParameterReinjectedInCreateNextRequestMethod.cs]: I agree, I was actually thinking the same but I thought I might be overcomplicating it. I should I have discussed it. Thanks for confirming though.

File: packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs [hunk 5]
```diff
@@ -123,6 +123,39 @@ public void AppendQueryDelimited<T>(string name, IEnumerable<T> value, string de
             AppendQuery(name, string.Join(delimiter, stringValues), escape);
         }
 
+        public void UpdateQuery(string name, string value)
+        {
+            if (PathAndQuery.Length == _pathLength)
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs:128]: If I'm understanding the logic here, if there are no query parameters, we add the new query parameter. Then, we check `PathAndQuery` again to see if it contains the query parameter which we just added, which it should. 

Could this not just be simplified to have the if / else you have below? If there are no query params, then it would go in the else block.

File: packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs [hunk 6]
```diff
@@ -123,6 +123,39 @@ public void AppendQueryDelimited<T>(string name, IEnumerable<T> value, string de
...
+            {
+                PathAndQuery.Append('?');
+                PathAndQuery.Append(name);
+                PathAndQuery.Append('=');
+                PathAndQuery.Append(value);
+            }
+            if (PathAndQuery.ToString().Contains(name + "="))
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs]: I'm wondering if all of this could be simplified to use the StringBuilder [Replace](https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.replace?view=net-10.0) method? That would avoid having to build the string. To be safe, we could ensure we only ever replace the first occurrence by passing a start + end index?

File: packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs [hunk 7]
```diff
@@ -123,6 +123,39 @@ public void AppendQueryDelimited<T>(string name, IEnumerable<T> value, string de
...
+                {
+                    valueEndIndex = currentQuery.Length;
+                }
+                string newQuery = currentQuery.Substring(0, valueStartIndex) + value + currentQuery.Substring(valueEndIndex);
+                PathAndQuery.Remove(_pathLength + 1, PathAndQuery.Length - _pathLength - 1);
+                PathAndQuery.Append(newQuery);
+            }
+            else
+            {
+                PathAndQuery.Append('&');
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs]: any reason why we didn't just use the `AppenQuery` overload?

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/ClientUriBuilderTests.cs [hunk 8]
```diff
@@ -481,5 +481,21 @@ public void AppendQueryDelimited(string endpoint, IEnumerable<string> value, str
...
             Assert.AreEqual(expected, builder.ToUri().ToString());
         }
+
+        [TestCase("http://localhost", "param1", "value1", "http://localhost/?param1=value1")]
+        [TestCase("http://localhost?existing=old", "param1", "value1", "http://localhost/?existing=old&param1=value1")]
+        [TestCase("http://localhost?param1=old", "param1", "new", "http://localhost/?param1=new")]
+        [TestCase("http://localhost?param1=old&param2=value2", "param1", "new", "http://localhost/?param1=new&param2=value2")]
+        [TestCase("http://localhost?param2=value2&param1=old", "param1", "new", "http://localhost/?param2=value2&param1=new")]
+        [TestCase("http://localhost?param2=value2&param1=old&param3=value3", "param1", "new", "http://localhost/?param2=value2&param1=new&param3=value3")]
+        public void UpdateQuery(string endpoint, string name, string value, string expected)
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/test/Providers/ClientUriBuilderTests.cs:493]: This is a great start! Could we add an edge case where another query parameter _contains_ a substring of what we are replacing? ie.

`[TestCase("http://localhost?fooparam1=value2&param1=old&param3=value3", "param1", "new", "http://localhost/?fooparam1=value2&param1=new&param3=value3")]`

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/RestClientProvider.cs [hunk 9]
```diff
@@ -460,14 +469,29 @@ private static List<MethodBodyStatement> AppendQueryParameters(ScopedApi uri, In
             return statement;
         }
 
+        private static MethodBodyStatement BuildUpdateQueryStatement(
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/RestClientProvider.cs]: nit: There's some code duplication here. There's an opportunity to consolidate some of the functionality here with `BuildAppendQueryStatement` 😃

File: packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs [hunk 10]
```diff
@@ -123,6 +123,53 @@ public void AppendQueryDelimited<T>(string name, IEnumerable<T> value, string de
             AppendQuery(name, string.Join(delimiter, stringValues), escape);
         }
 
+        public void UpdateQuery(string name, string value)
+        {
+            if (PathAndQuery.Length == _pathLength)
+            {
+                PathAndQuery.Append('?');
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Sample-TypeSpec/src/Generated/Internal/ClientUriBuilder.cs]: could we simply call AppendQuery in this block as well?

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ClientUriBuilderDefinition.cs [hunk 11]
```diff
@@ -334,6 +336,84 @@ private MethodProvider BuildAppendDelimitedMethod(string appendDelimitedMethodNa
...
+        {
+            var nameParameter = new ParameterProvider("name", $"The name.", typeof(string));
+            var valueParameter = new ParameterProvider("value", $"The value.", typeof(string));
+
+            var signature = new MethodSignature(
+                Name: UpdateQueryMethodName,
+                Modifiers: MethodSignatureModifiers.Public,
+                Parameters: [nameParameter, valueParameter],
+                ReturnType: null,
+                Description: null, ReturnDescription: null);
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ClientUriBuilderDefinition.cs]: nit: can we put ReturnDescription in a new line as well?

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ClientUriBuilderDefinition.cs [hunk 12]
```diff
@@ -334,6 +336,84 @@ private MethodProvider BuildAppendDelimitedMethod(string appendDelimitedMethodNa
...
+
+            var stringBuilder = PathAndQueryProperty.As<StringBuilder>();
+            var pathLength = (ValueExpression)_pathLengthField;
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ClientUriBuilderDefinition.cs]: we can probably just do `_pathLengthField.AsValueExpression` to be explicit.

**Review by jorgerangel-msft — APPROVED** [review]
